### PR TITLE
CO-1925 simplify code and perform an extra combinations (all payments)

### PR DESCRIPTION
**Pull request by nbornand**

### CODE REVIEW

File: sponsorship_compassion/models/account_invoice.py [hunk 1]
```diff
@@ -124,65 +124,53 @@ def _group_or_split_reconcile(self):
...
-                            matching_lines += open_payments[prev_index:][index]
-                            prev_index += index + 1
-                        break
-                if matching_lines:
-                    break
+            all_combinations = \
+                (combination for n in range(2, len(open_payments) + 1)
+                 for combination in itertools.combinations(open_payments, n))
+            for combination in all_combinations:
+                total = sum(payment.credit for payment in combination)
```
Comment (ecino) [inline sponsorship_compassion/models/account_invoice.py]: Even better : `total = sum(combination.mapped('credit'))`. This becomes more clear if `combination` is renamed to `payments_combination`

File: sponsorship_compassion/models/account_invoice.py [hunk 2]
```diff
@@ -124,65 +124,53 @@ def _group_or_split_reconcile(self):
...
+                if total == reconcile_amount:
+                    for payment in combination:
```
Comment (ecino) [inline sponsorship_compassion/models/account_invoice.py]: Why not directly do `(combination | move_lines).reconcile()`

**Review by ecino — CHANGES_REQUESTED** [review]
I think it's getting better, just small propositions.

File: sponsorship_compassion/models/account_invoice.py (hunk 2 above)
Comment (nbornand) [reply sponsorship_compassion/models/account_invoice.py]: I guessed it was there for a reason. Don't both operands have to be RecordSets?

File: sponsorship_compassion/models/account_invoice.py (hunk 1 above)
Comment (nbornand) [reply sponsorship_compassion/models/account_invoice.py]: not sure the mapped works as itertools.combinations yields tupples

File: sponsorship_compassion/models/account_invoice.py [hunk 3]
```diff
@@ -150,13 +150,13 @@ def _group_or_split_reconcile(self):
...
                 (combination for n in range(2, len(open_payments) + 1)
                  for combination in itertools.combinations(open_payments, n))
-            for combination in all_combinations:
-                total = sum(payment.credit for payment in combination)
-                if total == reconcile_amount:
-                    for payment in combination:
+            for payment_combination in all_payment_combinations:
+                combination_amount = sum(p.credit for p in payment_combination)
+                if combination_amount == reconcile_amount:
+                    for payment in payment_combination:
```
Comment (ecino) [inline sponsorship_compassion/models/account_invoice.py]: You can avoid this loop by doing directly `(payment_combination | move_lines).reconcile()`

**Review by ecino — CHANGES_REQUESTED** [review]

File: sponsorship_compassion/models/account_invoice.py (hunk 1 above)
Comment (ecino) [reply sponsorship_compassion/models/account_invoice.py]: Ah you're right, I didn't see these are tuples.

File: sponsorship_compassion/models/account_invoice.py (hunk 2 above)
Comment (ecino) [reply sponsorship_compassion/models/account_invoice.py]: It was my same mistaking, I didn't see combination is a tuple and not a recordset. So I think it's good now.

**Review by ecino — APPROVED** [review]

**Comment by ecino** [comment]
@nbornand Did you try a similar case we looked at to see if behaviour still looks good?

**Pull request merged**